Maple2.Server.Game/Util/DamageCalculator.cs

```diff
@@ -0,0 +1,146 @@
+                caster.Stats.Values[BasicAttribute.Accuracy].Total, caster.Buffs.GetResistance(BasicAttribute.Evasion), caster.Stats.Values[BasicAttribute.Luck].Total)) {
+            return (DamageType.Miss, 0);
+        }
+
+        var damageType = DamageType.Normal;
+        (double minBonusAtkDamage, double maxBonusAtkDamage) = caster.Stats.GetBonusAttack(target.Buffs.GetResistance(BasicAttribute.BonusAtk), target.Buffs.GetResistance(BasicAttribute.MaxWeaponAtk));
+        double attackDamage = minBonusAtkDamage + (maxBonusAtkDamage - minBonusAtkDamage) * Random.Shared.NextDouble();
+
+        // change the NPCNormalDamage to be changed depending on target?
+        double damageBonus = 1 + (caster.Stats.Values[SpecialAttribute.TotalDamage].Multiplier()) + (caster.Stats.Values[SpecialAttribute.NormalNpcDamage].Multiplier());
+        if (target is FieldNpc npc && npc.Value.IsBoss) {
+            damageBonus += caster.Stats.Values[SpecialAttribute.BossNpcDamage].Multiplier();
+        }
+
+        // Get elemental bonus
+        double elementBonus = 0;
+        switch (record?.SkillMetadata.Property.Element) {
+            case Element.Fire:
+                elementBonus = caster.Stats.Values[SpecialAttribute.FireDamage].Multiplier();
+                break;
+            case Element.Ice:
+                elementBonus = caster.Stats.Values[SpecialAttribute.IceDamage].Multiplier();
+                break;
+            case Element.Electric:
+                elementBonus = caster.Stats.Values[SpecialAttribute.ElectricDamage].Multiplier();
+                break;
+            case Element.Holy:
+                elementBonus = caster.Stats.Values[SpecialAttribute.HolyDamage].Multiplier();
+                break;
+            case Element.Dark:
+                elementBonus = caster.Stats.Values[SpecialAttribute.DarkDamage].Multiplier();
+                break;
+            case Element.Poison:
+                elementBonus = caster.Stats.Values[SpecialAttribute.PoisonDamage].Multiplier();
+                break;
+        }
+        damageBonus += elementBonus;
+
+        // Get range bonus
+        switch (record?.SkillMetadata.Property.RangeType) {
+            case RangeType.None:
+                break;
+            case RangeType.Melee:
+                damageBonus += caster.Stats.Values[SpecialAttribute.MeleeDamage].Multiplier();
+                break;
+            case RangeType.Range:
+                damageBonus += caster.Stats.Values[SpecialAttribute.RangedDamage].Multiplier();
+                break;
+        }
+
+        damageBonus += -target.Buffs.GetResistance(BasicAttribute.AttackSpeed) * caster.Stats.Values[BasicAttribute.AttackSpeed].Total;
+
+        // Check for crit and get crit damage
+        if (!dotDamage && record != null && record.AttackMetadata.CompulsionTypes.Contains(CompulsionType.Critical)) {
```

Comment:
Dots can crit based on their useGrade. Both dots and skill hits can be guaranteed crits from soaring focus tonic, sin's fatal strikes, and specific skills having compulsionType 2.